File: apollo-http-cache/src/main/java/com/apollographql/apollo/cache/http/CacheResponseBody.java
package com.apollographql.apollo.cache.http;

import okhttp3.MediaType;
import okhttp3.ResponseBody;
import okio.BufferedSource;
import okio.Okio;
import okio.Source;

final class CacheResponseBody extends ResponseBody {
  private BufferedSource responseBodySource;
  private final String contentType;
  private final String contentLength;

  CacheResponseBody(Source responseBodySource, String contentType, String contentLength) {
    this.responseBodySource = Okio.buffer(responseBodySource);
    this.contentType = contentType;
    this.contentLength = contentLength;
  }

  @Override public MediaType contentType() {
    return contentType != null ? MediaType.parse(contentType) : null;
  }

  @Override public long contentLength() {
    try {
      return contentLength != null ? Long.parseLong(contentLength) : -1;
    } catch (NumberFormatException e) {
      return -1;
    }
  }

  @Override public BufferedSource source() {
    return responseBodySource;
  }
}

File: apollo-http-cache/src/main/kotlin/com/apollographql/apollo3/cache/http/CacheResponseBody.kt
package com.apollographql.apollo3.cache.http

import okhttp3.MediaType
import okhttp3.ResponseBody
import okio.BufferedSource
import okio.Source
import okio.buffer

internal class CacheResponseBody(responseBodySource: Source, contentType: String?, contentLength: String?) : ResponseBody() {
  private val responseBodySource: BufferedSource
  private val contentType: String?
  private val contentLength: String?
  override fun contentType(): MediaType? {
    return if (contentType != null) MediaType.parse(contentType) else null
  }

  override fun contentLength(): Long {
    return try {
      contentLength?.toLong() ?: -1
    } catch (e: NumberFormatException) {
      -1
    }
  }

  override fun source(): BufferedSource {
    return responseBodySource
  }

  init {
    this.responseBodySource = responseBodySource.buffer()
    this.contentType = contentType
    this.contentLength = contentLength
  }
}